Site Editor: Add 'Show template' toggle when editing pages

noisysocks opened the pull request:
## What?
First pass at adding a _Show template_ toggle to the site editor when editing a page. Toggling this hides the blocks belonging to the template entirely so that you're left with a minimal interface that mimics the post editor.

## Why?
Allows users to focus on their content when desired. Moreover, it is small a step towards one day unifying the post editor and site editor. See item 3 in https://github.com/WordPress/gutenberg/issues/41717.

## How?
We re-use the approach that the site editor has for editing a `wp_navigation` CPT but instead of creating a fake `core/navigation` block we create a fake `core/group` block containing a `core/post-title, `core/post-content` and `core/post-featured-image`. These blocks are shown in the same order they appear in the template and do not appear if the template doesn't have it.

## Testing Instructions

1. Go to Appearance → Editor → Pages, select or create a page, then edit it.
2. In the Page settings sidebar, click on the Template dropdown and toggle "Show template" and "Hide template"
3. Check that the page content blocks are the same between hide/show modes.
4. Hide the template and then navigate away from edit mode in the site editor by clicking on the WP logo. The template should reappear in view mode.
5. Navigate to another section, e.g., Patterns, then return to the same page you edited previously
6. The page blocks should appear as expected

Run some tests:

```
npm run test:unit packages/edit-site/src/components/block-editor/block-editor-provider/test/use-page-content-blocks.js
```

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/6458278/41792555-9c99-4b5e-86a2-24638352f361

noisysocks committed:
Site Editor: First pass at adding a Show Template toggle when editing pages

noisysocks committed:
Bit of clean up

ramonjd commented:
I know this is a proposal, so just commenting on the concept of having an icon (any icon) in the Document tools toolbar to toggle between editing a page/post and editing that page/post's template.

This is all totally subjective. 

I'm finding it hard to articulate why I'm not sold on it yet.

Though it seems like the right place for such a button, given that its siblings are mostly related to document editing, the toolbar does contain a number of tools already, it might be just that I don't like clutter 🤷🏻 

Or could be the icon itself that's throwing me. Maybe a lock/unlock set would tap my brain in the right way and it's just as simple as finding the right signage.

With this toggle button, we'd have three different controls to toggle between modes (counting the notification), whose functions require the user to expend a few cognitive cycles to either read their descriptions or learn what they do. This is not a bad thing. The editor is full of such things.

I mention this with the risk of bringing up a potentially tired pattern, but I can't help but wonder if the editor couldn't just  show me straight up what I can and cannot edit in the canvas.

Example:

<img width="1267" alt="Screenshot 2023-07-19 at 12 29 37 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/d45d748c-6dd1-4a12-88b4-386ee2d91dd6">

jameskoster commented:
Did y'all consider an inverted implementation of the button, IE use the `layout` icon toggled on when the template is visible? I'm wondering if that might be more intuitive... the Post Content icon doesn't really communicate _post content_ all that well.

~Personally I wouldn't include the post title when focussed on post content as it is technically part of the layout. For me the toggle here is essentially; view all blocks versus view contents of Post Content block. I appreciate this means you wouldn't be able to edit the title when focussed on the content, but we need to add an off-canvas title editing UI for post formats anyway. There should be a 'Rename' action in the Command Palette too.~

Edit: From discussions elsewhere it seems like this isn't strictly about toggling the template. It's more about hiding parts of the template that distract from writing (headers, footers, sidebars, etc.). It can still be useful for certain meta data (title, featured image) to be editable on the canvas though.

noisysocks commented:
> Did y'all consider an inverted implementation of the button, IE use the `layout` icon toggled on when the template is visible? I'm wondering if that might be more intuitive... the Post Content icon doesn't really communicate _post content_ all that well.

We didn't spend much time thinking about which way the toggle should work as it's easy to swap 😀 Agree the `layout` icon is more comprehensible.

noisysocks committed:
Use a toggle in sidebar

noisysocks committed:
Make toggling 'Show template' on/off faster

noisysocks committed:
Use blocks in same order as template

noisysocks committed:
Tidy up code

noisysocks commented:
I updated this so that _Show template_ is a toggle that appears in the sidebar. This is a lot less prominent which makes me feel comfortable shipping it in the plugin as a way to gather feedback and iterate. I also refactored things so that switching between modes is fast. I'm pretty happy with it, so will mark it as ready for review, though maybe @SaxonF wants to push a few design tweaks.

noisysocks committed:
Add unit tests

noisysocks committed:
Add doc comment

jameskoster commented:
Given the changes to the template section of the sidebar (https://github.com/WordPress/gutenberg/pull/51477) we might do something like this:

<img width="279" alt="Screenshot 2023-09-12 at 17 24 47" src="https://github.com/WordPress/gutenberg/assets/846565/8180ee23-ba0d-4ee8-bc55-4a7c97c71480">

It has similar prominence to the current state of the PR.

ramonjd commented:
I'll put this on my list to revive. Thanks for the updated designs @jameskoster 🍺

ramonjd commented on packages/edit-site/src/components/sidebar-edit-mode/page-panels/edit-template.js:
```diff
@@ -83,6 +93,20 @@ export default function EditTemplate() {
+										? 'hideTemplate'
+										: 'disableTemplate'
+								);
+								onClose();
```
Do folks think we need any `info` help text for this menu item or are the labels adequate?

ramonjd committed:
Formatting
Remove unused component import
Doc update

ramonjd commented on docs/reference-guides/data/data-core-edit-site.md:
```diff
@@ -295,7 +312,7 @@ _Returns_
 ### setHasPageContentFocus
 
-Sets whether or not the editor allows only page content to be edited.
+Sets whether the editor allows only page content to be edited.
```
Grammatical change only

jameskoster commented:
Thanks for the update @ramonjd. 

A small suggestion in terms of flow: Instead of a button with adaptive label ('show template' / 'hide template'), could we try a single checked menu item? That way you can toggle the template on and off without having to re-open the popover. Basically the same behavior as editor preferences like "Top toolbar".

Another observation: there's a small chance the user might think this option affects the frontend too. Wondering what we can do to eliminate any possible confusion there. Some options:

<img width="1296" alt="Screenshot 2023-09-14 at 10 52 05" src="https://github.com/WordPress/gutenberg/assets/846565/37f0a54e-8d70-469c-bbe4-a4610021add7">

Coupled with the behavior requested above, I think the "Template preview" label might be adequate. Any thoughts @SaxonF?

ramonjd commented:
The dropdown's width adjusts with the content: 

https://github.com/WordPress/gutenberg/assets/6458278/9258d25d-30e5-44e5-a322-e5de1cc343ff

We could mitigate this slightly by adding an "off" icon, e.g., 

https://github.com/WordPress/gutenberg/assets/6458278/afbad92b-da94-4b6a-8fd7-b29a56c52c4d


Here it is with some info:

https://github.com/WordPress/gutenberg/assets/6458278/681080d2-7b52-4d2c-a984-05633a82d72f


In my mind one of the first two is simpler. "Template preview" implies that it's not a destructive change. Or something like "Preview template in the editor" 🤷🏻 

Happy to commit whatever folks think is best 😄

ramonjd committed:
With icon and updated copy

andrewserong commented:
This is testing quite nicely for me!

> The dropdown's width adjusts with the content:

If we don't want to add a `reset` icon for the off state, could we add a `min-width` to the `.components-popover__content` to work around the issue? Something like the following seems to work okay for me locally:

```scss
.edit-site-page-panels-edit-template__dropdown {
	.components-popover__content {
		min-width: 240px;
	}
}
```

Quick demo:

https://github.com/WordPress/gutenberg/assets/14988353/009d6714-f3ff-45ef-b64a-9efa34cc9abe

I added it to `sidebar-edit-mode/page-panels/style.scss` and I see there's already a similar min-width rule for another popover, so maybe it'd be okay?: https://github.com/WordPress/gutenberg/blob/bba4bbf8c998432159c6b1ea7db40bfe4fc23d5d/packages/edit-site/src/components/sidebar-edit-mode/page-panels/style.scss#L37-L41

jameskoster commented:
Yeah the min-width solution seems good to me.

ramonjd committed:
Use only the check icon and set min width of dropdown to minimize the effect of the dropdown width toggling between narrow and wide

ramonjd commented:
> could we add a min-width to the .components-popover__content to work around the issue? Something like the following seems to work okay for me locally:

Thanks @andrewserong and @jameskoster 

I did consider it, but stopped short at the thought that we might not be able to cater for particularly verbose translations. Here's a contrived preview:

https://github.com/WordPress/gutenberg/assets/6458278/d1cc3fd0-820a-4df0-a62d-e0194c27d134

But it should cover most cases so I've updated the PR with this suggestion.

Cheers!

andrewserong reviewed (commented):
Thanks @ramonjd! The toggle is looking good now 👍

While re-testing, I think I've encountered a bug if you switch the Template preview off, go out of the page, and then go back into it via the menu at the top left, without reloading the page. What seems to happen is that the preview mode blocks wind up being duplicated:

https://github.com/WordPress/gutenberg/assets/14988353/7ab7e7bc-54fe-48c0-af88-517d83c57e18

I'm not _too_ sure what's causing it, but I wonder if `usePageGhostBlocks` winds up calling `createBlock` too often / not re-using the existing blocks?

Semi-related, but looking at this also made me wonder about what the desired behaviour is when users navigate away from the page that they're editing. If we reload the page, then it defaults to being switched off, which I'd expect. But if a user navigates to global styles, or away from the page via the button at the top left, then we currently maintain the preview mode. I could imagine folks potentially getting a little stuck and forgetting where they toggled the value.

Design-wise, would it be worth adding an icon next to the Template button to indicate when the template preview is switched off? Alternately, should we reset the state for the preview when folks navigate to one of the browse-mode menus?

ramonjd commented:
> go out of the page, and then go back into it via the menu at the top left, without reloading the page. What seems to happen is that the preview mode blocks wind up being duplicated:

Oh! I'll have a deeper look and try to find out why. 

> Design-wise, would it be worth adding an icon next to the Template button to indicate when the template preview is switched off? Alternately, should we reset the state for the preview when folks navigate to one of the browse-mode menus?

Good ideas. I like the idea of the former, but it might be difficult to communicate via icon. Creating a consistent "leave the page" experience — the preview state is reset to whatever the default is — sounds to me like a better approach in the short term. 

Thanks for testing this so thoroughly 🙇🏻

jameskoster commented:
> Creating a consistent "leave the page" experience — the preview state is reset to whatever the default is — sounds to me like a better approach in the short term.

That sounds good to me. 

In the future we might say that posts toggle the template preview off by default while everything else toggles on. There can also be site settings to adjust those default states. 

I don't think we need an icon to indicate the off state, the canvas should make it obvious in most cases.

andrewserong commented:
> I don't think we need an icon to indicate the off state, the canvas should make it obvious in most cases.

Agreed — if we're resetting the state when leaving the page (going back to the menu), then I think that'll nicely account for the case that I was imagining of a user navigating away and then forgetting how to get back to displaying the template preview. Sounds good!

ramonjd committed:
This commit tries to get around a peculiarity of __experimentalGetGlobalBlocksByName, which doesn't seem to be able to find the post-content block in a new block tree (if it even exists in state).
We're sidestepping the issue, but with fewer selectors, so overall hopefully it's more sustainable.

ramonjd commented:
> go out of the page, and then go back into it via the menu at the top left, without reloading the page. What seems to happen is that the preview mode blocks wind up being duplicated:

As far as I can tell, when [usePageGhostBlocks](https://github.com/WordPress/gutenberg/blob/c5fa9ff93bd93f5673cda4e9737aee15ae3f2aa7/packages/edit-site/src/components/block-editor/block-editor-provider/default-block-editor-provider.js#L72) looks up the current
 `PAGE_CONTENT_BLOCK_TYPES` it's finding those that are _currently_ rendered in the editor. 

So when we navigate from the page back to the pages menu in the sidebar, the template changes to the home page in the editor. 

That means the block list has changed, and therefore the following selectors are now trying to find  `PAGE_CONTENT_BLOCK_TYPES` in the home page template. 

I confirmed this by matching the returned clientIds.

So in the home page template, `PAGE_CONTENT_BLOCK_TYPES`  may or may not be there. 

For example, there is no `core/post-content` block in my home page template, but there are `core/post-title` and `core/post-featured-image`.

https://github.com/WordPress/gutenberg/assets/6458278/932d6dba-9aa1-4d4e-a4d6-877fc02420db

And because `__experimentalGetGlobalBlocksByName`returns all global blocks that match a blockName including nested blocks, they may be "found" twice, reflecting what @andrewserong reported [above](https://github.com/WordPress/gutenberg/pull/52674#pullrequestreview-1630207078).

```js
		return getBlockNamesByClientId(
			__experimentalGetGlobalBlocksByName( PAGE_CONTENT_BLOCK_TYPES )
		);
```

The fundamental issue I believe is that `__experimentalGetGlobalBlocksByName`, for some reason, can't find the post-content block in the block list state - if it even exists at all.

I don't know why?

Rather than spend all day debugging the store, in 69deb7d10ebf9d0a32893345f36776929fef1d15 I've side-stepped the issue, and am filtering the content blocks from the block list that's actually rendered.

> if we're resetting the state when leaving the page (going back to the menu), then I think that'll nicely account for the case that I was imagining of a user navigating away and then forgetting how to get back to displaying the template preview.

I haven't implemented this yet, hoping to ensure we've covered the first bug first 🤣 

I was thinking that we could achieve the reset in the site-editor package by flipping the `pageContentFocusType` state when `getCanvasMode()` returns `view`.

andrewserong commented on packages/edit-site/src/components/block-editor/block-editor-provider/default-block-editor-provider.js:
```diff
@@ -0,0 +1,118 @@
+						},
+					},
+				},
+				flattenBlocks( blocks ).filter( ( block ) =>
+					PAGE_CONTENT_BLOCK_TYPES.includes( block.name )
+				)
```
Tiny nit: this is probably not a big deal, especially since this is only executed within a `useMemo`, but we could potentially avoid having to iterate over the `blocks` again for a `filter()` call if in `flattenBlocks` we switched to a while loop / stack-based approach as in the block editor reducer here:

https://github.com/WordPress/gutenberg/blob/21b6a3717c1b09c376f7a5d914b0953aa2cdcddd/packages/block-editor/src/store/reducer.js#L80-L91

In that approach, we'd push to `result` as we encounter blocks, rather than having to filter after the fact. We could also add the block name to the result list instead of the whole object in case that simplifies things, as well as potentially skip iterating into children if we encounter a Query Loop block?

Another thought about filtering the existing blocks vs calling `createBlock` with the block name is that filtering existing blocks might also carry over styling that could exist at the block level in the template, which may or may not be desired when template preview is switched off. I wasn't sure which would be preferred 🤔

ramonjd replied:
> In the commits prior to https://github.com/WordPress/gutenberg/commit/69deb7d10ebf9d0a32893345f36776929fef1d15, I think we were only creating one of each block, matched by name?

Oh yeah, in my testing I forgot to revert back to createBlock for one of each page types. 😅 

I'll take another look based on your research, thanks again!

Good to hear that it's actually rendering something useful now though!

ramonjd replied:
I've reverted back to using `createBlock`. Also the while loop/stack has a funny side effect of reversing the order of found objects even using `pop()`, so I've opted for a basic, but reliable for loop + continue. 

I think it works better, and have tested it with `core/query` as well as multiples of page content blocks. I'm going to write some unit tests tomorrow to increase my confidence. Right now, you are my unit tests and that's not fair on you 🤣

andrewserong replied:
> Also the while loop/stack has a funny side effect of reversing the order of found objects even using pop(), so I've opted for a basic, but reliable for loop + continue.

Ah, it possibly would need `stack.push` to be `stack.unshift` instead for it to be a depth-first traversal, otherwise siblings might be accessed before inner blocks as inner blocks get placed on the end of the stack? No matter, though, the `for` loop with `continue` looks fine to me 🙂

andrewserong replied:
> Right now, you are my unit tests and that's not fair on you 🤣

Always happy to be a unit test, the only problem is that I'm non-deterministic 😆

ramonjd replied:
> otherwise siblings might be accessed before inner blocks as inner blocks get placed on the end of the stack?

At that point I thought it might have created for both me and others a bit of an order-related mind cluster bomb, so I thought the for loop was neater for the human brain 😄

andrewserong replied:
> so I thought the for loop was neater for the human brain 

Totally, better not to include "code test"-style examples / brain teasers in the code base unless necessary. `for` loops are much easier to read IMO, too 🙂

andrewserong commented on packages/edit-site/src/components/block-editor/block-editor-provider/default-block-editor-provider.js:
```diff
@@ -0,0 +1,118 @@
+	return (
+		<ExperimentalBlockEditorProvider
+			settings={ settings }
+			value={ isTemplateHidden ? pageGhostBlocks : blocks }
```
Should we also check that `pageGhostBlocks` isn't empty? It's technically possible for `pageGhostBlocks` to be an empty array, so if it is empty, would we want to pass `blocks` instead? 🤔

andrewserong reviewed (commented):
Nice investigation @ramonjd! That seems to get us quite a bit closer, and resolves going in and out of the same page for me. While testing, I noticed another issue where we can unexpectedly encounter duplicate blocks, though it's for a different reason.

In one of my custom templates, I have a Query Loop block that contains Title and Featured Image blocks — the use case is a header or footer area of a tiled list of posts. It seems the logic in this PR will match against the Title, Featured Image or Content blocks, irrespective of whether or not they're within a Query Loop block.

Here's a screengrab of loading a page with such a template, editing the template, then switching back to the page edit mode, and switching off the template preview, revealing the duplicate Featured Image block:

https://github.com/WordPress/gutenberg/assets/14988353/98e3e0e0-a55e-4042-96f3-8183269555a6

In the commits prior to https://github.com/WordPress/gutenberg/pull/52674/commits/69deb7d10ebf9d0a32893345f36776929fef1d15, I think we were only creating one of each block, matched by name?

I've left a comment where we might be able to introduce skipping iterating into children of the Query Loop block, but feel free to take a different approach if you have other ideas!

> I haven't implemented this yet, hoping to ensure we've covered the first bug first 🤣

Good thinking!

ramonjd committed:
Removing the filter iteration layer and adding it to the flattenBlock helper function
Checking for disallowed blocks as well, e.g., core/query

ramonjd commented on packages/edit-site/src/components/block-editor/block-editor-provider/default-block-editor-provider.js:
```diff
@@ -0,0 +1,135 @@
+	);
+}
+
+const identity = ( x ) => x;
```
Probably can ditch this since it's only used once internally and always with a `transform()`

ramonjd commented on packages/edit-site/src/components/block-editor/block-editor-provider/default-block-editor-provider.js:
```diff
@@ -0,0 +1,135 @@
+ * of the template and its nested entities.
+ *
+ * If the page content focus type is `'hideTemplate'`, the provider will provide
+ * a set of "ghost" blocks that mimick the look and feel of the post editor and
```
Note to self: update comments to remove reference to "ghost" 👻 and take about page content view or something

ramonjd commented on packages/edit-site/src/components/block-editor/block-editor-provider/default-block-editor-provider.js:
```diff
@@ -0,0 +1,135 @@
+ *
+ * @return {Array} Page content blocks.
+ */
+function usePageContentBlocks( blocks, isPageContentFocused = false ) {
```
TODO: Unit tests for this hook. They'll indirectly cover `flattenBlocks` as well

andrewserong commented on packages/edit-site/src/components/block-editor/block-editor-provider/default-block-editor-provider.js:
```diff
@@ -0,0 +1,135 @@
+					},
+				},
+				flattenBlocks( blocks, ( block ) => {
+					if ( PAGE_CONTENT_BLOCK_TYPES.includes( block.name ) ) {
```
Another tiny optimisation idea, again feel free to ignore! Would it be worth making `PAGE_CONTENT_BLOCK_TYPES` an object with key / value pairs for block name, with the value being `true`, that way you could access `PAGE_CONTENT_BLOCK_TYPES[ block.name ]` instead of needing to call `.includes()`?

ramonjd replied:
Oh yes, why not?! I was just blindly following Rob's stuff. I think it's only used in a couple of places so shouldn't add too much beef to this PR. Cheers

apeatling reviewed (commented):
Looks like you guys have everything well covered. I had a look at this and confirming that it worked as expected when creating an editing a new page in the site editor:

![2023-09-19 13 24 35](https://github.com/WordPress/gutenberg/assets/1464705/84314927-fe05-4419-a6cc-de1dceeea3ec)

ramonjd commented:
> I had a look at this and confirming that it worked as expected when creating an editing a new page in the site editor:

Thanks for testing @apeatling Hoping that the alert popup isn't caused by this PR 😬 😆

apeatling commented:
> Thanks for testing @apeatling Hoping that the alert popup isn't caused by this PR 😬 😆

Gifox needs a Sonoma update :D

ramonjd committed:
Extract usePageContentBlocks and write basic tests
Moving PAGE_CONTENT_BLOCK_TYPES to utils/constants and creating a map

andrewserong commented on packages/edit-site/src/store/actions.js:
```diff
@@ -676,3 +676,20 @@ export const toggleDistractionFree =
+ * - `'disableTemplate'`: Disable the blocks belonging to the page's template.
+ * - `'hideTemplate'`: Hide the blocks belonging to the page's template.
+ *
+ * @param {'disbleTemplate'|'hideTemplate'} pageContentFocusType The type of page content focus.
```
```suggestion
 * @param {'disableTemplate'|'hideTemplate'} pageContentFocusType The type of page content focus.
```

andrewserong commented on packages/edit-site/src/store/actions.js:
```diff
@@ -676,3 +676,20 @@ export const toggleDistractionFree =
+ *
+ * @return {Object} Action object.
+ */
+export function setPageContentFocusType( pageContentFocusType ) {
+	return {
+		type: 'SET_PAGE_CONTENT_FOCUS_TYPE',
+		pageContentFocusType,
+	};
+}
```
This might be getting a little nit-picky, so my apologies if this isn't useful feedback! Begin thoughts:

Just looking over this again, and I'm wondering a bit about the addition of `setPageContentFocusType` and `getPageContentFocusType` and the potential overlap between these and `setHasPageContentFocus`. Since the latter is already a non-private action, I'm not sure how well we can consolidate things, but I'm wondering if it might be possible?

My main concern is that `setPageContentFocusType`'s state is pretty much dependent on `setPageContentFocus` being `true`, so depending on the circumstances it might be awkward having both bits of state.

Just thinking out loud, but would it be possible to consolidate via something like:

* Update `hasPageContentFocus` selector and reducer to look at the same piece of state as `getPageContentFocusType`, but the `hasPageContentFocus` just checks that the value is truthy.
* Update `setPageContentFocus` so that it's string-based, but if it accepts a `boolean` then `true` maps to `disableTemplate`?

I think that could group things together, though it might not be desirable if we ever wish to persist the state of the focus type separately to whether or not page content focus is set. If that's the case then feel free to disregard this feedback!

On the other hand, if we're undecided about which direction we'd like to go in the future, then another option could be to move the action and selector added in this PR to be private.

ramonjd replied:
With @noisysocks gone it's hard to gauge the intention behind `pageContentFocusType` — maybe other "types" were envisioned? 

I agree it seems to be pretty tightly coupled with `hasPageContentFocus`.

Thanks for thinking about an alternative. I think we could take a different approach as you mentioned, but I'm not convinced that expanding the return type of an already public selector with a bunch of usages will make things easier to work with in the future.

Also, as you mention there could be a case in which the states are independent. Locking a template while editing a page vs hiding a template while editing a page seem to me to be two separate actions.

> another option could be to move the action and selector added in this PR to be private

I was actually in the middle of doing this, so I think our instincts are in alignment insofar as the ambiguity of the selector 😄

andrewserong replied:
> I'm not convinced that expanding the return type of an already public selector with a bunch of usages will make things easier to work with in the future.

Very good point! Best not to mess with the public signature 👍

> I was actually in the middle of doing this, so I think our instincts are in alignment insofar as the ambiguity of the selector 😄

Nice one! Sounds good 🙂

andrewserong commented on packages/edit-site/src/components/block-editor/block-editor-provider/test/use-page-content-blocks.js:
```diff
@@ -0,0 +1,83 @@
+	];
+	it( 'should return empty array if `isPageContentFocused` is `false`', () => {
+		const { result } = renderHook( () =>
+			usePageContentBlocks( blocksList, false )
```
Tiny nit: feel free to disregard!

Would it be clearer if `usePageContentBlocks` accepted an object instead of positional params? That way the boolean in the second param could be clearer as to what the value refers to.

On the other hand, since it is a boolean, it makes for a nice on / off switch irrespective of what the source of the boolean is... so I don't mind it after all. Ignore me! 😄

ramonjd replied:
Fair comment! I also thought about whether the args were appropriate but shelved it as a refactor job since it's only local to the component.

andrewserong replied:
Absolutely, I'd leave as-is since it's local as mention 👍

andrewserong commented on packages/edit-site/src/components/block-editor/block-editor-provider/test/use-page-content-blocks.js:
```diff
@@ -0,0 +1,83 @@
+			createBlock( 'core/query' ),
+			createBlock( 'core/post-content' ),
+		] ),
+		createBlock( 'core/query' ),
```
Should we add a test were there are blocks as children of `core/query` to check that it doesn't iterate into children of that block type?

ramonjd replied:
This is a great idea. Will do!

andrewserong reviewed (commented):
This is testing nicely for me now @ramonjd when I switch between pages, and when I go to a page that has content blocks within a query loop block 👍

Just left a few comments, some of them a bit nit-picky or rambling, so feel free to ignore any that aren't relevant. I reckon we're probably ready to add in the behaviour to reset the template preview mode when going out of / into individual pages within browse mode now, if you do?

ramonjd committed:
Updates selectors to include canvas mode so we can show any hidden templates in focus mode

ramonjd committed:
Moving getPageContentFocusType and setPageContentFocusType selector/action to locked, private methods
In the use-page-content-blocks.js test, add innerblocks to the query test block

ramonjd commented:
> Creating a consistent "leave the page" experience — the preview state is reset to whatever the default is 

This functionality _should_ now be activated. 😄 🤞🏻

andrewserong commented on packages/edit-site/src/store/actions.js:
```diff
@@ -285,7 +285,7 @@ export const setPage =
 							per_page: -1,
 						} )
-				)?.find( ( { slug } ) => slug === currentTemplateSlug );
+				 )?.find( ( { slug } ) => slug === currentTemplateSlug );
```
I think prettier is complaining about the whitespace here 😄

<img width="693" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/faf0c11f-73c8-4714-9eca-5ab292ac094f">

ramonjd replied:
Thanks. 👍🏻 I saw that, and just reverted all the selector/actions changes which weren't relevant - already pushed. 

For some reason my IDE formatted it and it passed the pre-commit 🤷🏻 🤷🏻🤷🏻🤷🏻🤷🏻🤷🏻🤷🏻

andrewserong commented on packages/edit-site/src/components/block-editor/block-editor-provider/use-page-content-blocks.js:
```diff
@@ -0,0 +1,77 @@
+		if ( ! isPageContentFocused || ! blocks || ! blocks.length ) {
+			return [];
+		}
+		return [
+			createBlock(
+				'core/group',
+				{
+					layout: { type: 'constrained' },
+					style: {
+						spacing: {
+							margin: {
+								top: '4em', // Mimics the post editor.
+							},
+						},
+					},
+				},
+				flattenBlocks( blocks, ( block ) => {
+					if ( PAGE_CONTENT_BLOCK_TYPES[ block.name ] ) {
+						return createBlock( block.name );
+					}
+				} )
+			),
+		];
```
Another nit: if no page content blocks are found, then we return an empty group block. This means that if a page's template has no content blocks at all, switching off "Template preview" results in an empty page, because the return value is not an empty array:

<img width="1275" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/5a5b075a-9d28-45a5-a273-9e813d1d2c03">

This is getting very edge-casey territory, and you'd only see this if you go to a page that uses a template that has no content blocks and only after switching off Template preview, so not a blocker to landing for me!

ramonjd replied:
Great find! I think we should follow up with a fix for this. 

I'll make a note and get a PR up. 

🙇🏻

andrewserong reviewed (approved):
Thanks again for all the rounds of follow-up @ramonjd, this is testing great for me now! The resetting when going to the browse mode is working nicely, with the value persisting when I go back into the editor. This feels like a good balance to me, so that folks going out and in to a page have their value persisted, but if they reload the page they can easily get back to the template preview, too 👍

https://github.com/WordPress/gutenberg/assets/14988353/d6f92b61-5e61-4e63-a83e-24b14f822093

I noted one other (very) edge case, which I've added in a comment — feel free to disregard that one for this PR, I think we're well and truly in polishing territory which could be looked at in a follow-up if need be. And of course we can see if we get feedback from folks using the feature during the 6.4 beta, too. For now this PR is feeling good and solid to me and ready to land 🎉

Great work again getting this PR to a thorough state and adding in all the tests. LGTM! ✨

ramonjd commented:
@andrewserong Thanks for sticking with me on this one. I owe you 💯 🍹

ramonjd commented:
Also found a bug. When exiting the site editor in edit mode with the template hidden, we need to reset `pageContentFocusType` to 'disabledState' otherwise it'll be hidden by default. Looking now.

youknowriad commented on packages/edit-site/src/components/block-editor/block-editor-provider/use-page-content-blocks.js:
```diff
@@ -0,0 +1,77 @@
+						},
+					},
+				},
+				flattenBlocks( blocks, ( block ) => {
```
I find it weird that we're trying to guess the blocks from the content of the template rather than just fixing "core/post-title" + "core/post-content" (like the post editor).

ramonjd replied:
Could you help me understand that statement? Is there something more reliable that could be done here?

I imagine it was approached this way to cover the (possibly unlikely) scenario that there are 2 x site titles or whatever in a template. 

Plus there is a bit of guessing required since we don't know how deep or where these blocks are located in a template in advance.

andrewserong replied:
> rather than just fixing "core/post-title" + "core/post-content" (like the post editor).

By grabbing the real blocks we ensure that the order of the content blocks matches the template so that it more closely resembles the site. I suppose the main value at the moment is that it determines whether or not a featured image is shown, and the order of the title, featured image, and content blocks. While reviewing, I thought it added a nice level of polish.

youknowriad replied:
> By grabbing the real blocks we ensure that the order of the content blocks matches the template so that it more closely resembles the site

Is that the role of this mode though? I mean, it felt like this mode is more of a "post editor" mode to me where we don't really care about the template, we're just filling the content and title.

ramonjd replied:
> I mean, it felt like this mode is more of a "post editor" mode to me where we don't really care about the template, we're just filling the content and title.

We could do it this way. 

I'd probably want to confer on the UX with some design folks.

The thinking was that, because we're in the site editor, it would be a more intuitive, seamless experience to have the page content reflect the page template.

So for instance, if I edit the Pages template to have the post title appear underneath the post content, (just because I can) then when I edit that page in the site editor, the page corresponds with the new template structure. I can see my changes. 

It might be confusing to have the page content seemingly 'revert' in template preview mode just seconds after I've edited the template, whereas the post editor doesn't have any of the extra context of templates etc so the experience is more "static", if that makes sense.

That's the argument for. 

Against,  it'd be easier to refactor the code to just make it match the post editor behavior, e.g., just show the required inner blocks...

```js
				[
					createBlock( 'core/post-title' ),
					createBlock( 'core/post-content' ),
				]
```

Another angle is to remove the template preview mode altogether, I'd be for this. I understand there's a "[great unification](https://github.com/WordPress/gutenberg/issues/41717)" going on, where the site editor will be jack of all trades, but I don't know how much value hiding the template in this way adds. Maybe to some folks.

andrewserong replied:
> I don't know how much value hiding the template in this way adds. Maybe to some folks.

I'm not sure how much value it has in its current form, I but I like the principle behind the idea of a more focused writing-mode like experience akin to the post editor, that folks can use from within the site editor. However it's exposed, I think it's important that folks can still set a featured image while editing their post content, whether or not that's in the sidebar or on the editor canvas. One of the neat things about having it in the canvas is that it gets exposed in the list view for free, so it's easier to find / interact with.

youknowriad reviewed (commented):
I'm in a middle of a big refactor to use the EntityProvider and would have appreciated an e2e test to ensure I didn't break this feature. #56000

ramonjd commented:
> I'm in a middle of a big refactor to use the EntityProvider and would have appreciated an e2e test to ensure I didn't break this feature. https://github.com/WordPress/gutenberg/pull/56000

I'll add one. Thanks!
